File: web-ui/src/api/intelligence.ts
/**
 * Frontend API client for backend-computed intelligence endpoints.
 *
 * These functions call the Rust backend which owns all derived/business
 * logic. The frontend passes only transient SSE data (permissions,
 * questions, watcher status, signals) and the backend merges it with
 * its own persisted state to compute results.
 */

import { apiFetch, apiPost } from "./client";

// ── Shared input types (match Rust serde shapes) ─────────

export interface PermissionInput {
  id: string;
  sessionID: string;
  toolName: string;
  description?: string;
  time: number;
}

export interface QuestionInput {
  id: string;
  sessionID: string;
  title: string;
  time: number;
}

export interface SignalInput {
  id: string;
  kind: string;
  title: string;
  body: string;
  created_at: number;
  session_id?: string | null;
}

export interface WatcherStatusInput {
  session_id: string;
  action: string;
  idle_since_secs?: number;
}

// ── Inbox ────────────────────────────────────────────────

export type InboxItemPriority = "high" | "medium" | "low";
export type InboxItemState = "unresolved" | "informational";
export type InboxItemSource = "permission" | "question" | "mission" | "watcher" | "completion";

export interface InboxItem {
  id: string;
  source: InboxItemSource;
  title: string;
  description: string;
  priority: InboxItemPriority;
  state: InboxItemState;
  created_at: number;
  session_id?: string | null;
  mission_id?: string | null;
}

export interface InboxResponse {
  items: InboxItem[];
}

export async function computeInbox(req: {
  permissions: PermissionInput[];
  questions: QuestionInput[];
  watcher_status?: WatcherStatusInput | null;
  signals: SignalInput[];
}): Promise<InboxResponse> {
  return apiPost<InboxResponse>("/inbox", req);
}

// ── Recommendations ──────────────────────────────────────

export type RecommendationAction =
  | "open_inbox"
  | "open_missions"
  | "open_memory"
  | "open_routines"
  | "open_delegation"
  | "open_workspaces"
  | "open_autonomy"
  | "setup_daily_summary"
  | "upgrade_autonomy_nudge"
  | "setup_daily_copilot";

export interface AssistantRecommendation {
  id: string;
  title: string;
  rationale: string;
  action: RecommendationAction;
  priority: InboxItemPriority;
}

export interface RecommendationsResponse {
  recommendations: AssistantRecommendation[];
}

export async function computeRecommendations(req: {
  permissions: PermissionInput[];
  questions: QuestionInput[];
}): Promise<RecommendationsResponse> {
  return apiPost<RecommendationsResponse>("/recommendations", req);
}

// ── Handoffs ─────────────────────────────────────────────

export interface HandoffLink {
  kind: string;
  label: string;
  source_id?: string | null;
}

export interface HandoffBrief {
  title: string;
  summary: string;
  blockers: string[];
  recent_changes: string[];
  next_action: string;
  links: HandoffLink[];
}

export async function computeMissionHandoff(req: {
  mission_id: string;
  permissions: PermissionInput[];
  questions: QuestionInput[];
}): Promise<HandoffBrief | null> {
  try {
    return await apiPost<HandoffBrief>("/handoff/mission", req);
  } catch {
    return null;
  }
}

export async function computeSessionHandoff(req: {
  session_id: string;
  permissions: PermissionInput[];
  questions: QuestionInput[];
}): Promise<HandoffBrief | null> {
  try {
    return await apiPost<HandoffBrief>("/handoff/session", req);
  } catch {
    return null;
  }
}

// ── Resume Briefing ──────────────────────────────────────

export interface ResumeBriefing {
  title: string;
  summary: string;
  next_action: string;
}

export async function computeResumeBriefing(req: {
  active_session_id?: string | null;
  permissions: PermissionInput[];
  questions: QuestionInput[];
  signals: SignalInput[];
}): Promise<ResumeBriefing | null> {
  try {
    return await apiPost<ResumeBriefing>("/resume-briefing", req);
  } catch {
    return null;
  }
}

// ── Daily Summary ────────────────────────────────────────

export interface DailySummaryResponse {
  summary: string;
}

export async function computeDailySummary(req: {
  routine_id: string;
  permissions: PermissionInput[];
  questions: QuestionInput[];
  signals: SignalInput[];
}): Promise<DailySummaryResponse> {
  return apiPost<DailySummaryResponse>("/daily-summary", req);
}

// ── Signals ──────────────────────────────────────────────

export interface SignalsResponse {
  signals: SignalInput[];
}

export async function fetchSignals(): Promise<SignalsResponse> {
  return apiFetch<SignalsResponse>("/signals");
}

export async function addSignal(req: {
  kind: string;
  title: string;
  body: string;
  session_id?: string | null;
}): Promise<SignalInput> {
  return apiPost<SignalInput>("/signals", req);
}

// ── Assistant Center Stats ───────────────────────────────

export interface AssistantCenterStats {
  active_missions: number;
  blocked_missions: number;
  total_missions: number;
  pending_permissions: number;
  pending_questions: number;
  memory_items: number;
  active_routines: number;
  active_delegations: number;
  workspace_count: number;
  autonomy_mode: string;
}

export async function computeAssistantStats(req: {
  permissions: PermissionInput[];
  questions: QuestionInput[];
}): Promise<AssistantCenterStats> {
  return apiPost<AssistantCenterStats>("/assistant-center/stats", req);
}

// ── Workspace Templates ──────────────────────────────────

export interface WorkspaceTemplate {
  id: string;
  name: string;
  description: string;
  panels: { sidebar: boolean; terminal: boolean; editor: boolean; git: boolean };
  layout: { sidebar_width: number; terminal_height: number; side_panel_width: number };
}

export interface WorkspaceTemplatesResponse {
  templates: WorkspaceTemplate[];
}

export async function fetchWorkspaceTemplates(): Promise<WorkspaceTemplatesResponse> {
  return apiFetch<WorkspaceTemplatesResponse>("/workspace-templates");
}

// ── Active Memory ────────────────────────────────────────

export interface ActiveMemoryResponse {
  memory: Array<{
    id: string;
    label: string;
    content: string;
    scope: string;
    project_index?: number | null;
    session_id?: string | null;
    created_at: string;
    updated_at: string;
  }>;
}

export async function fetchActiveMemory(
  projectIndex?: number,
  sessionId?: string | null,
): Promise<ActiveMemoryResponse> {
  const params = new URLSearchParams();
  if (projectIndex !== undefined) params.set("project_index", String(projectIndex));
  if (sessionId) params.set("session_id", sessionId);
  const qs = params.toString();
  return apiFetch<ActiveMemoryResponse>(`/memory/active${qs ? `?${qs}` : ""}`);
}

// ── Routine Summary (backend-computed) ───────────────────

export async function computeRoutineSummary(req: {
  routine_id: string;
  active_session_id?: string | null;
  permissions?: PermissionInput[];
  questions?: QuestionInput[];
  signals?: SignalInput[];
}): Promise<DailySummaryResponse> {
  return apiPost<DailySummaryResponse>("/daily-summary", {
    routine_id: req.routine_id,
    permissions: req.permissions ?? [],
    questions: req.questions ?? [],
    signals: req.signals ?? [],
  });
}

File: leptos-ui/src/api/intelligence.rs
//! Intelligence/Assistant API — matches React `api/intelligence.ts`.

use serde::Serialize;
use crate::types::api::{
    InboxResponse, RecommendationsResponse, AssistantCenterStats,
};
use super::client::{api_fetch, api_post, ApiError};

// ── Request types ───────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct PermissionInput {
    pub id: String,
    pub session_id: String,
    pub tool: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub args_summary: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct QuestionInput {
    pub id: String,
    pub session_id: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct SignalInput {
    pub id: String,
    pub kind: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct WatcherStatusInput {
    pub session_id: String,
    pub action: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub idle_since_secs: Option<u64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct InboxComputeRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub permissions: Option<Vec<PermissionInput>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub questions: Option<Vec<QuestionInput>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signals: Option<Vec<SignalInput>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub watchers: Option<Vec<WatcherStatusInput>>,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecommendationsRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_index: Option<usize>,
}

#[derive(Debug, Clone, Serialize)]
pub struct HandoffRequest {
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ResumeBriefingRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_index: Option<usize>,
}

#[derive(Debug, Clone, Serialize)]
pub struct DailySummaryRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_index: Option<usize>,
}

#[derive(Debug, Clone, Serialize)]
pub struct AddSignalRequest {
    pub kind: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct AssistantStatsRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_index: Option<usize>,
}

// ── Response types ──────────────────────────────────────────────────

#[derive(Debug, Clone, serde::Deserialize)]
pub struct HandoffLink {
    pub label: String,
    pub url: String,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct HandoffBrief {
    pub session_id: String,
    pub summary: String,
    pub context: String,
    pub next_steps: Vec<String>,
    pub links: Vec<HandoffLink>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct ResumeBriefing {
    pub summary: String,
    pub active_sessions: Vec<String>,
    pub pending_items: Vec<String>,
    pub suggested_next: Vec<String>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct DailySummaryResponse {
    pub summary: String,
    pub highlights: Vec<String>,
    pub metrics: serde_json::Value,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct SignalEntry {
    pub id: String,
    pub kind: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
    pub timestamp: String,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct SignalsResponse {
    pub signals: Vec<SignalEntry>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct WorkspaceTemplate {
    pub name: String,
    pub description: String,
    pub panels: serde_json::Value,
    pub layout: serde_json::Value,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct WorkspaceTemplatesResponse {
    pub templates: Vec<WorkspaceTemplate>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct ActiveMemoryResponse {
    pub items: Vec<crate::types::api::PersonalMemoryItem>,
}

// ── API functions ───────────────────────────────────────────────────

/// Compute inbox items (merging permissions, questions, signals, watchers).
pub async fn compute_inbox(req: &InboxComputeRequest) -> Result<InboxResponse, ApiError> {
    api_post("/inbox", req).await
}

/// Compute assistant recommendations.
pub async fn compute_recommendations(req: &RecommendationsRequest) -> Result<RecommendationsResponse, ApiError> {
    api_post("/recommendations", req).await
}

/// Compute session handoff briefing.
pub async fn compute_session_handoff(req: &HandoffRequest) -> Result<HandoffBrief, ApiError> {
    api_post("/handoff/session", req).await
}

/// Compute resume briefing.
pub async fn compute_resume_briefing(req: &ResumeBriefingRequest) -> Result<ResumeBriefing, ApiError> {
    api_post("/resume-briefing", req).await
}

/// Compute daily summary.
pub async fn compute_daily_summary(req: &DailySummaryRequest) -> Result<DailySummaryResponse, ApiError> {
    api_post("/daily-summary", req).await
}

/// Fetch signals.
pub async fn fetch_signals() -> Result<SignalsResponse, ApiError> {
    api_fetch("/signals").await
}

/// Add a signal.
pub async fn add_signal(req: &AddSignalRequest) -> Result<SignalEntry, ApiError> {
    api_post("/signals", req).await
}

/// Compute assistant center stats.
pub async fn compute_assistant_stats(req: &AssistantStatsRequest) -> Result<AssistantCenterStats, ApiError> {
    api_post("/assistant-center/stats", req).await
}

/// Fetch workspace templates.
pub async fn fetch_workspace_templates() -> Result<Vec<WorkspaceTemplate>, ApiError> {
    let resp: WorkspaceTemplatesResponse = api_fetch("/workspace-templates").await?;
    Ok(resp.templates)
}

/// Fetch active/scoped memory.
pub async fn fetch_active_memory(
    project_index: Option<usize>,
    session_id: Option<&str>,
) -> Result<Vec<crate::types::api::PersonalMemoryItem>, ApiError> {
    let mut path = "/memory/active".to_string();
    let mut sep = '?';
    if let Some(idx) = project_index {
        path.push_str(&format!("{}project_index={}", sep, idx));
        sep = '&';
    }
    if let Some(sid) = session_id {
        path.push_str(&format!("{}session_id={}", sep, js_sys::encode_uri_component(sid)));
    }
    let resp: ActiveMemoryResponse = api_fetch(&path).await?;
    Ok(resp.items)
}

/// Compute routine summary (alias for daily-summary with routine focus).
pub async fn compute_routine_summary(req: &DailySummaryRequest) -> Result<DailySummaryResponse, ApiError> {
    api_post("/daily-summary", req).await
}
